mjpc/planners/cem/planner.cc

```diff
@@ -0,0 +1,527 @@
+  }
+
+  winner = 0;
+}
+
+// allocate memory
+void CEMPlanner::Allocate() {
+  // initial state
+  int num_state = model->nq + model->nv + model->na;
+
+  // state
+  state.resize(num_state);
+  mocap.resize(7 * model->nmocap);
+  userdata.resize(model->nuserdata);
+
+  // policy
+  int num_max_parameter = model->nu * kMaxTrajectoryHorizon;
+  policy.Allocate(model, *task, kMaxTrajectoryHorizon);
+  previous_policy.Allocate(model, *task, kMaxTrajectoryHorizon);
+
+  // scratch
+  parameters_scratch.resize(num_max_parameter);
+  times_scratch.resize(kMaxTrajectoryHorizon);
+
+  // noise
+  noise.resize(kMaxTrajectory * (model->nu * kMaxTrajectoryHorizon));
+  stdev.resize(kMaxTrajectory * (model->nu * kMaxTrajectoryHorizon));
+  fill(stdev.begin(), stdev.end(), noise_exploration);  // TODO: if this gets called before Initialize, then move this to above
+
+  // trajectory and parameters
+  winner = -1;
+  for (int i = 0; i < kMaxTrajectory; i++) {
+    trajectory[i].Initialize(num_state, model->nu, task->num_residual,
+                             task->num_trace, kMaxTrajectoryHorizon);
+    trajectory[i].Allocate(kMaxTrajectoryHorizon);
+    candidate_policy[i].Allocate(model, *task, kMaxTrajectoryHorizon);
+  }
+
+  // need to initialize an arbitrary order of the trajectories
+  trajectory_order.reserve(num_trajectory_);
+  for (int i = 0; i < num_trajectory_; i++) {
+    trajectory_order.push_back(i);
+  }
+
+  // noise gradient
+  noise_gradient.resize(num_max_parameter);
+}
+
+// reset memory to zeros
+void CEMPlanner::Reset(int horizon) {
+  // state
+  std::fill(state.begin(), state.end(), 0.0);
+  std::fill(mocap.begin(), mocap.end(), 0.0);
+  std::fill(userdata.begin(), userdata.end(), 0.0);
+  time = 0.0;
+
+  // policy parameters
+  policy.Reset(horizon);
+  previous_policy.Reset(horizon);
+
+  // scratch
+  std::fill(parameters_scratch.begin(), parameters_scratch.end(), 0.0);
+  std::fill(times_scratch.begin(), times_scratch.end(), 0.0);
+
+  // noise
+  std::fill(noise.begin(), noise.end(), 0.0);
+
+  // trajectory samples
+  for (int i = 0; i < kMaxTrajectory; i++) {
+    trajectory[i].Reset(kMaxTrajectoryHorizon);
+    candidate_policy[i].Reset(horizon);
+  }
+
+  for (const auto& d : data_) {
+    mju_zero(d->ctrl, model->nu);
+  }
+
+  // noise gradient
+  std::fill(noise_gradient.begin(), noise_gradient.end(), 0.0);
+
+  // improvement
+  improvement = 0.0;
+
+  // winner
+  winner = 0;
+}
+
+// set state
+void CEMPlanner::SetState(const State& state) {
+  state.CopyTo(this->state.data(), this->mocap.data(), this->userdata.data(),
+               &this->time);
+}
+
+int CEMPlanner::OptimizePolicyCandidates(int ncandidates, int horizon,
+                                              ThreadPool& pool) {
+  // if num_trajectory_ has changed, use it in this new iteration.
+  // num_trajectory_ might change while this function runs. Keep it constant
+  // for the duration of this function.
+  int num_trajectory = num_trajectory_;
```

Comment:
you might want to recompute n_elite here since num_trajectory_ can be changed in the GUI

Reply:
I think it's OK - `n_elites` gets passed in as the first argument and doesn't intrinsically depend on `num_trajectory_` (that's only used for the initial value right now).